eth/downloader/downloader.go

```diff
@@ -1001,223 +909,191 @@ func (d *Downloader) fetchHeaders(p *peerConnection, from uint64) error {
-		p.log.Trace("Fetching next pivot header", "number", pivot+uint64(fsMinFullBlocks))
-		go p.peer.RequestHeadersByNumber(pivot+uint64(fsMinFullBlocks), 2, fsMinFullBlocks-9, false) // move +64 when it's 2x64-8 deep
-	}
-	// Start pulling the header chain skeleton until all is done
-	ancestor := from
-	getHeaders(from)
+		case errCanceled:
+			// Sync cancelled, no issue, propagate up
+			return err
 
-	mode := d.getMode()
-	for {
-		select {
-		case <-d.cancelCh:
-			return errCanceled
+		default:
+			// Header retrieval either timed out, or the peer failed in some strange way
+			// (e.g. disconnect). Consider the master peer bad and drop
+			d.dropPeer(p.id)
 
-		case packet := <-d.headerCh:
-			// Make sure the active peer is giving us the skeleton headers
-			if packet.PeerId() != p.id {
-				log.Debug("Received skeleton from incorrect peer", "peer", packet.PeerId())
-				break
+			// Finish the sync gracefully instead of dumping the gathered data though
+			for _, ch := range []chan bool{d.queue.blockWakeCh, d.queue.receiptWakeCh} {
+				select {
+				case ch <- false:
+				case <-d.cancelCh:
+				}
 			}
-			headerReqTimer.UpdateSince(request)
-			timeout.Stop()
-
-			// If the pivot is being checked, move if it became stale and run the real retrieval
-			var pivot uint64
-
-			d.pivotLock.RLock()
-			if d.pivotHeader != nil {
-				pivot = d.pivotHeader.Number.Uint64()
+			select {
+			case d.headerProcCh <- nil:
+			case <-d.cancelCh:
 			}
-			d.pivotLock.RUnlock()
+			return fmt.Errorf("%w: header request failed: %v", errBadPeer, err)
+		}
+		// If the pivot is being checked, move if it became stale and run the real retrieval
+		var pivot uint64
 
-			if pivoting {
-				if packet.Items() == 2 {
-					// Retrieve the headers and do some sanity checks, just in case
-					headers := packet.(*headerPack).headers
+		d.pivotLock.RLock()
+		if d.pivotHeader != nil {
+			pivot = d.pivotHeader.Number.Uint64()
+		}
+		d.pivotLock.RUnlock()
 
-					if have, want := headers[0].Number.Uint64(), pivot+uint64(fsMinFullBlocks); have != want {
-						log.Warn("Peer sent invalid next pivot", "have", have, "want", want)
-						return fmt.Errorf("%w: next pivot number %d != requested %d", errInvalidChain, have, want)
-					}
-					if have, want := headers[1].Number.Uint64(), pivot+2*uint64(fsMinFullBlocks)-8; have != want {
-						log.Warn("Peer sent invalid pivot confirmer", "have", have, "want", want)
-						return fmt.Errorf("%w: next pivot confirmer number %d != requested %d", errInvalidChain, have, want)
-					}
-					log.Warn("Pivot seemingly stale, moving", "old", pivot, "new", headers[0].Number)
-					pivot = headers[0].Number.Uint64()
+		if pivoting {
+			if len(headers) == 2 {
+				if have, want := headers[0].Number.Uint64(), pivot+uint64(fsMinFullBlocks); have != want {
+					log.Warn("Peer sent invalid next pivot", "have", have, "want", want)
+					return fmt.Errorf("%w: next pivot number %d != requested %d", errInvalidChain, have, want)
+				}
```

Comment:
We have these various spread-out checks "did X deliver what we asked"? 
Right now, the dispatcher guarantees that 
- A response is matched with a request, 
- The response is of the correct type, and can be cast:ed to the expected response type. 

However, the early checks could do more: 
- For headers, it could verify the amount, numbers/hashes against the request, 
- For bodies, it could verify the sha of the body

Feels like a lot of the clauses here would go away if we did that, and maybe we could move some processing off to be multithreaded.